### PR TITLE
Scheduler Test Cases

### PR DESCRIPTION
**Pull request by CaitieM20**
Adding Test Cases to the Scheduler

- Property based testing to ensure a variety of jobs will be ran and executed successfully
- Unit tests to validate errors are handled appropriately and tasks not executed when failure happens
- refactored the RunTask part of RunJob into its own function

### CODE REVIEW

File: sched/generators.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		Def: jobDef,
+	}
+
+	return job
+}
+
+func GenTask() TaskDefinition {
+	return TaskDefinition{
+		Command: Command{
+			Argv: []string{"cmd1", "cmd2"},
```
Comment (dbentley) [inline sched/generators.go]: FWIW, I think these would be more like "cmd", "arg1".

File: sched/generators.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		},
+	}
+}
+
+// Randomly generates an Id that is valid for
+// use as a sagaId or taskId
+func GenJobId() gopter.Gen {
+	return func(genParams *gopter.GenParameters) *gopter.GenResult {
```
Comment (dbentley) [inline sched/generators.go]: It's super-annoying that gopter.gen package wants you to work with monads in a language that's so hostile (syntactically, and type-system-wise) to monads... : /

File: sched/scheduler/jobRunner.go [hunk 3]
```diff
@@ -82,6 +60,32 @@ func runJob(job sched.Job, saga *s.Saga, nodes []cm.Node) {
 	return
 }
 
+// Logic to execute a single task in a Job.  Ensures Messages are logged to SagaLog
+// Logs StartTask, executes Tasks, Logs EndTask
+func runTask(saga *s.Saga, node cm.Node, taskId string, task sched.TaskDefinition) {
+	// Put StartTask Message on SagaLog
+	stErr := saga.StartTask(taskId, nil)
```
Comment (dbentley) [inline sched/scheduler/jobRunner.go]: why have different err's instead of just reusing err?

File: sched/scheduler/jobRunner.go [hunk 4]
```diff
@@ -82,6 +60,32 @@ func runJob(job sched.Job, saga *s.Saga, nodes []cm.Node) {
...
+	}
+
+	// TODO: After a number of attempts we should stop
+	// Trying to run a task, could be a poison pill
+	// Implement deadletter queue
+	taskExecuted := false
+	for !taskExecuted {
+		execErr := node.SendMessage(task)
+		if execErr == nil {
+			taskExecuted = true
```
Comment (dbentley) [inline sched/scheduler/jobRunner.go]: if you just break here you can skip having taskExecuted variable. If you want to keep it, I'd make it a 3-clause for loop (with an empty increment, for now)

File: sched/scheduler/scheduler_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+			cluster, clusterState := ci.DynamicLocalNodeClusterFactory(int(numNodes))
+			sagaCoord := saga.MakeInMemorySagaCoordinator()
+			scheduler := NewScheduler(cluster, clusterState, sagaCoord)
+
+			job := sched.GenJob(jobId, int(numTasks))
+			err := scheduler.ScheduleJob(job)
+
+			if err != nil {
+				fmt.Println("UnExpected Error Scheduling Job")
```
Comment (dbentley) [inline sched/scheduler/scheduler_test.go]: s/UnE/Une/

**Comment by dbentley** [comment]
LGTM

File: sched/generators.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
+		length := int(genParams.NextUint64()%20) + 1
+		result := make([]byte, length)
+		for i := 0; i < length; i++ {
+			result[i] = chars[genParams.Rng.Intn(len(chars))]
```
Comment (jasonbs10) [inline sched/generators.go]: What's the different between nextUint64 and rng.Intn?

File: sched/generators.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+	job := Job{
+		Id:  id,
+		Def: jobDef,
+	}
+
+	return job
+}
+
+func GenTask() TaskDefinition {
```
Comment (jasonbs10) [inline sched/generators.go]: Is this too deep to generate for property tests?

**Comment by jasonbs10** [comment]
lgtm

File: sched/generators.go (hunk 1 above)
Comment (CaitieM20) [inline sched/generators.go]: cool fixed!  We can make these more realistic or random if that'd be useful for worker testing in the future too!

File: sched/generators.go (hunk 2 above)
Comment (CaitieM20) [inline sched/generators.go]: yea its a Quick Check port which was originally written in Haskell and Erlang so its sort of from that world.  I agree it gets a little ugly sometimes

File: sched/scheduler/jobRunner.go (hunk 4 above)
Comment (CaitieM20) [inline sched/scheduler/jobRunner.go]: why have the empty increment?  the golang Tour of Go shows this pattern when you want a while loop? https://tour.golang.org/flowcontrol/3

File: sched/scheduler/scheduler_test.go (hunk 5 above)
Comment (CaitieM20) [inline sched/scheduler/scheduler_test.go]: fixed!

File: sched/generators.go (hunk 6 above)
Comment (CaitieM20) [inline sched/generators.go]: I don't think anything, will use rng.Intn much cleaner!

File: sched/generators.go (hunk 7 above)
Comment (CaitieM20) [inline sched/generators.go]: I think as the values applied to tasks gets used more, i.e. worker code gets implemented, then we can definitely make this more random as a generator.  Or we start doing datalocality based scheduling.

Comment (CaitieM20) [inline sched/generators.go]: I definitely see this code evolving over time and generating real workloads

**Pull request merged**